File: src/python/algorithms/math/maxAreaRectangle/max_area_rectangle.py
# HashSet + diagonal pair enumeration — TC: O(n²)  SC: O(n)
def max_area_rectangle(points):
    point_set = set()
    for x, y in points:
        point_set.add((x, y))

    max_area = 0
    n = len(points)

    for i in range(n):
        for j in range(i + 1, n):
            x1, y1 = points[i]
            x2, y2 = points[j]

            # skip same row or column (no rectangle possible)
            if x1 == x2 or y1 == y2:
                continue

            # check if the other two corners exist
            if (x1, y2) in point_set and (x2, y1) in point_set:
                area = abs(x1 - x2) * abs(y1 - y2)
                max_area = max(max_area, area)

    return max_area


if __name__ == "__main__":
    points1 = [[1,1],[1,3],[3,1],[3,3],[2,2]]
    r1 = max_area_rectangle(points1)
    print(f"Test 1 - corners at (1,1)(3,3): {'[PASS]' if r1 == 4 else '[FAIL]'} result={r1} expected=4")

    points2 = [[1,1],[1,3],[3,1],[3,3],[4,1],[4,3]]
    r2 = max_area_rectangle(points2)
    print(f"Test 2 - max rect is 3x2: {'[PASS]' if r2 == 6 else '[FAIL]'} result={r2} expected=6")

    points3 = [[0,0],[0,1],[1,0],[1,1]]
    r3 = max_area_rectangle(points3)
    print(f"Test 3 - unit square: {'[PASS]' if r3 == 1 else '[FAIL]'} result={r3} expected=1")

    points4 = [[0,0],[0,2],[3,0],[3,2]]
    r4 = max_area_rectangle(points4)
    print(f"Test 4 - 3x2 rectangle: {'[PASS]' if r4 == 6 else '[FAIL]'} result={r4} expected=6")

    points5 = [[1,1],[2,2],[3,1],[1,3]]
    r5 = max_area_rectangle(points5)
    print(f"Test 5 - no valid rectangle: {'[PASS]' if r5 == 0 else '[FAIL]'} result={r5} expected=0")

File: src/javascript/algorithms/math/maxAreaRectangle/maxAreaRectangle.js
// HashSet + diagonal pair enumeration — TC: O(n²)  SC: O(n)
function maxAreaRectangle(points) {
    const pointSet = new Set(points.map(([x, y]) => `${x}#${y}`));
    const n = points.length;
    let maxArea = 0;

    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            const [x1, y1] = points[i];
            const [x2, y2] = points[j];

            // skip same row or column (no rectangle possible)
            if (x1 === x2 || y1 === y2) continue;

            // check if the other two corners exist
            if (pointSet.has(`${x1}#${y2}`) && pointSet.has(`${x2}#${y1}`)) {
                const area = Math.abs(x1 - x2) * Math.abs(y1 - y2);
                maxArea = Math.max(maxArea, area);
            }
        }
    }

    return maxArea;
}

const points1 = [[1,1],[1,3],[3,1],[3,3],[2,2]];
const r1 = maxAreaRectangle(points1);
console.log(`Test 1 - corners at (1,1)(3,3): ${r1 === 4 ? '[PASS]' : '[FAIL]'} result=${r1} expected=4`);

const points2 = [[1,1],[1,3],[3,1],[3,3],[4,1],[4,3]];
const r2 = maxAreaRectangle(points2);
console.log(`Test 2 - max rect is 3x2: ${r2 === 6 ? '[PASS]' : '[FAIL]'} result=${r2} expected=6`);

const points3 = [[0,0],[0,1],[1,0],[1,1]];
const r3 = maxAreaRectangle(points3);
console.log(`Test 3 - unit square: ${r3 === 1 ? '[PASS]' : '[FAIL]'} result=${r3} expected=1`);

const points4 = [[0,0],[0,2],[3,0],[3,2]];
const r4 = maxAreaRectangle(points4);
console.log(`Test 4 - 3x2 rectangle: ${r4 === 6 ? '[PASS]' : '[FAIL]'} result=${r4} expected=6`);

const points5 = [[1,1],[2,2],[3,1],[1,3]];
const r5 = maxAreaRectangle(points5);
console.log(`Test 5 - no valid rectangle: ${r5 === 0 ? '[PASS]' : '[FAIL]'} result=${r5} expected=0`);
